compression support; no context only

y3llowcake opened the pull request.

garyburd commented on client.go:
```diff
@@ -221,6 +231,12 @@ func (d *Dialer) Dial(urlStr string, requestHeader http.Header) (*Conn, *http.Re
+	if d.CompressionSupported {
+		req.Header.Set("Sec-Websocket-Extensions", "permessage-defalte; server_no_context_takeover; client_no_context_takeover")
+	} else {
+		delete(req.Header, "Sec-Websocket-Extensions")
```
Follow pattern for Sec-Websocket-Protocol above. The dialer should return an error instead of silently deleting the the application's header.

garyburd commented on server.go:
```diff
@@ -126,6 +132,24 @@ func (u *Upgrader) Upgrade(w http.ResponseWriter, r *http.Request, responseHeade
+	if responseHeader == nil {
+		responseHeader = http.Header{}
+	} else {
+		delete(responseHeader, "Sec-Websocket-Extensions")
```
Do not modify the application's headers.  Follow the pattern used for Sec-Websocket-Protocol.

garyburd commented on client.go:
```diff
@@ -70,6 +74,12 @@ type Dialer struct {
+
+	// CompressionSupported specifies if the client should attempt to negotiate per
+	// message compression (RFC 7692). Setting this value to true does not
+	// gaurentee that compression will be supported. Currently only "no context
```
s/gaurentee/guarantee/

garyburd commented:
The performance when compression is enabled is not very good and should be investigated after this goes in.  Until the performance issues are figured out, I'd like the section in the package comment to also state that the feature is experimental.

y3llowcake commented:
I've addressed your comments. It wasn't completely clear to me how you wanted to handle application provided Sec-Websocket-Extensions headers in the Upgrader. I chose to fail, as allowing the application to pass this header would likely result in broken behavior.

garyburd commented:
Avoid modifying the application's response header in Upgrader.Upgrade.

Check for the invalid response header near the beginning of the function:

     if _, ok := responseHeader["Sec-Websocket-Extensions"]; ok {
         return u.returnError(w, r, http.StatusInternalServerError, "websocket: application specific Sec-Websocket-Extensions headers are unsupported")
    }

Negotiate compression without scribbling on application's headier.

    var compress bool
    if u.CompressionSupported {
        for _, ext := range parseExtensions(r.Header) {
            if ext[""] != "permessage-deflate" {
                continue
            }
            compress = true
            break
        }
    }

Add the header directly to response:

    if compress {
        p = append(p, "Sec-Websocket-Extensions: permessage-deflate; server_no_context_takeover; client_no_context_takeover\r\n")
    }

y3llowcake commented:
dunzo

garyburd commented:
Looks good. Can you squash the changes?

FZambia commented:
Looks like there is [a typo](https://github.com/gorilla/websocket/pull/161/files#diff-d0b7f86cea13f90d3cc0cc2338da8d5fR234) - `permessage-defalte`

y3llowcake commented:
Good catch on the typo. Commits have been squashed.
